src/Browser/Test/HasBrowser.php

```diff
@@ -16,147 +11,15 @@
-            );
-
-            return new $class(self::$primaryPantherClient);
-        });
-
-        BrowserExtension::registerBrowser($browser);
-
-        return $browser
-            ->setSourceDir($_SERVER['BROWSER_SOURCE_DIR'] ?? './var/browser/source')
-            ->setScreenshotDir($_SERVER['BROWSER_SCREENSHOT_DIR'] ?? './var/browser/screenshots')
-            ->setConsoleLogDir($_SERVER['BROWSER_CONSOLE_LOG_DIR'] ?? './var/browser/console-logs')
-        ;
-    }
-
-    protected function httpBrowser(): HttpBrowser
-    {
-        $browser = HttpBrowser::create(function() {
-            $class = $_SERVER['HTTP_BROWSER_CLASS'] ?? HttpBrowser::class;
-
-            if (!\is_a($class, HttpBrowser::class, true)) {
-                throw new \RuntimeException(\sprintf('"HTTP_BROWSER_CLASS" env variable must reference a class that extends %s.', HttpBrowser::class));
-            }
-
-            $baseUri = $_SERVER['HTTP_BROWSER_URI'] ?? null;
-
-            if (!$baseUri && !$this instanceof PantherTestCase) {
-                throw new \RuntimeException(\sprintf('If not using "HTTP_BROWSER_URI", your TestCase must extend "%s".', PantherTestCase::class));
-            }
-
-            if (!$baseUri) {
-                self::startWebServer();
-
-                $baseUri = self::$baseUri;
-            }
-
-            // copied from PantherTestCaseTrait::createHttpBrowserClient()
-            $client = new HttpBrowserClient();
-            $urlComponents = \parse_url($baseUri);
-            $host = $urlComponents['host'];
-
-            if (isset($urlComponents['port'])) {
-                $host .= ":{$urlComponents['port']}";
-            }
-
-            $client->setServerParameter('HTTP_HOST', $host);
-
-            if ('https' === ($urlComponents['scheme'] ?? 'http')) {
-                $client->setServerParameter('HTTPS', 'true');
-            }
-
-            /** @var HttpBrowser $browser */
-            $browser = new $class(self::$httpBrowserClients[] = $client);
-
-            if (!$this instanceof KernelTestCase) {
-                return $browser;
-            }
-
-            if (!static::$booted) {
-                static::bootKernel();
-            }
-
-            if (static::$container->has('profiler')) {
-                $browser->setProfiler(static::$container->get('profiler'));
-            }
-
-            return $browser;
-        });
-
-        BrowserExtension::registerBrowser($browser);
-
-        return $browser
-            ->setSourceDir($_SERVER['BROWSER_SOURCE_DIR'] ?? './var/browser/source')
-        ;
+    use HasHttpBrowser, HasKernelBrowser, HasPantherBrowser {
+        HasKernelBrowser::browser insteadof HasHttpBrowser;
+        HasKernelBrowser::browser as kernelBrowser;
+        HasHttpBrowser::browser as httpBrowser;
+        HasPantherBrowser::browser as pantherBrowser;
     }
 
-    protected function kernelBrowser(): KernelBrowser
+    protected function browser(): void
```

Comment:
I'm not too familiar with "trait hacking" but doesn't the `insteadof` above prevent this from being called?

Reply:
I ran some tests and apparently it is not needed. See https://3v4l.org/RVre0 for a full test, some notes on "weird behavior":

* If when renaming all conflicted methods, we must add at least one `insteadof` statement
* We don't need to add more than one `insteadof` statement, eventhough more than 2 traits cause a conflict
* The class method always overrides any trait methods

Reply:
I see, what do you think of browser being an alias for kernelBrowser? I think this is the most common browser.

Reply:
Yes, I thought about it but wasn't sure. I don't have a strong opinion here. I decided not to do it to have 2 clear ways of specifying your browser: either you use a specific trait with a generic method, or use a generic trait with a specific method. Having a generic trait with a generic method seemed a bit confusing to me.

Having said that, I'm not 100% convinced of the name "kernelBrowser". It makes all sense if you look at the implementation, but I'm not sure if this specific implementation is well-known under BrowserKit users (i.e. Symfony users). Maybe dropping "kernelBrowser" completely in favor of "browser" (so having a browser and specific HttpBrowser/PantherBrowser) is better? (not sure at all - let's discuss outside this PR)

Reply:
> Having a generic trait with a generic method seemed a bit confusing to me.

Ok, let's leave as is. You're right, if you're using multiple different browsers in a TestCase, you probably want to know which one is which at a glance.

> Having said that, I'm not 100% convinced of the name "kernelBrowser"

Me either... Ryan and I have a little brainstorm on this and came up with nothing... Yes, let's discuss further in slack.